coordinator/src/funding_fee.rs

```diff
@@ -0,0 +1,363 @@
+
+    scheduler.start().await?;
+
+    Ok(())
+}
+
+/// Generate [`FundingFeeEvent`]s for all active positions.
+///
+/// When called, a [`FundingFeeEvent`] will be generated for an active position if:
+///
+/// - We can get a [`FundingRate`] that is at most 1 hour old from the DB.
+/// - We can get a BitMEX index price for the `end_date` of the [`FundingRate`].
+/// - There is no other [`FundingFeeEvent`] in the DB with the same `position_id` and `end_date`.
+/// - The position was created _before_ the `end_date` of the [`FundingRate`].
+///
+/// This function should be safe to retry. Retry should come in handy if the index price is
+/// not available.
+fn generate_funding_fee_events(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    index_price_source: IndexPriceSource,
+    auth_users_notifier: tokio::sync::mpsc::Sender<OrderbookMessage>,
+) -> Result<()> {
+    let mut conn = pool.get()?;
+
+    tracing::debug!("Generating funding fee events");
+
+    let funding_rate = db::funding_rates::get_funding_rate_charged_in_the_last_hour(&mut conn)?;
+
+    let funding_rate = match funding_rate {
+        Some(funding_rate) => funding_rate,
+        None => {
+            tracing::debug!("No current funding rate for this hour");
+            return Ok(());
+        }
+    };
+
+    // TODO: Funding rates should be specific to contract symbols.
+    let contract_symbol = ContractSymbol::BtcUsd;
+
+    let index_price = match index_price_source {
+        IndexPriceSource::Bitmex => block_in_place(move || {
+            let current_index_price =
+                get_bitmex_index_price(&contract_symbol, funding_rate.end_date())?;
+
+            anyhow::Ok(current_index_price)
+        })?,
+        IndexPriceSource::Fake => {
+            #[cfg(not(debug_assertions))]
+            compile_error!("Cannot use a fake index price in release mode");
+
+            dec!(50_000)
+        }
+    };
+
+    if index_price.is_zero() {
+        bail!("Cannot generate funding fee events with zero index price");
+    }
+
+    // We exclude active positions which were open after this funding period ended.
+    let positions = db::positions::Position::get_all_active_positions_open_before(
+        &mut conn,
+        funding_rate.end_date(),
+    )?;
+    for position in positions {
+        let amount = calculate_funding_fee(
+            position.quantity,
+            funding_rate.rate(),
+            index_price,
+            position.trader_direction,
+        );
+
+        if let Some(funding_fee_event) = db::funding_fee_events::insert(
+            &mut conn,
+            amount,
+            position.trader,
+            position.id,
+            funding_rate.end_date(),
+            index_price,
+            funding_rate.rate(),
+        )
+        .context("Failed to insert funding fee event")?
+        {
+            block_in_place(|| {
+                auth_users_notifier
+                    .blocking_send(OrderbookMessage::TraderMessage {
```

Comment:
Since we go to the database a few times I thought it would be easier to make it sync and use `block_in_place`. I can't say I know which approach is better.

Maybe we should look into using https://github.com/weiznich/diesel_async for the coordinator.